lasagne/layers/recurrent.py

```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+        self.W_hid_to_gates = T.concatenate(
+            [self.W_hid_to_resetgate, self.W_hid_to_updategate,
+             self.W_hid_to_cell], axis=1)
+
+        # Stack gate biases into a (4*num_units) vector
+        self.b_gates = T.concatenate(
+            [self.b_resetgate, self.b_updategate,
+             self.b_cell], axis=0)
+
+        # Initialize hidden state
+        if isinstance(hid_init, T.TensorVariable):
+            if hid_init.ndim != 2:
+                raise ValueError("When a tensor hid_init should be a matrix")
+            self.hid_init = hid_init
+        else:
+            self.hid_init = self.add_param(
+                hid_init, (1, self.num_units), name="hid_init",
+                trainable=learn_init, regularizable=False)
+
+    def get_output_shape_for(self, input_shape):
+        return input_shape[0], input_shape[1], self.num_units
+
+    def get_output_for(self, input, mask=None, *args, **kwargs):
+        """
+        Compute this layer's output function given a symbolic input variable
+
+        Parameters
+        ----------
+        input : theano.TensorType
+            Symbolic input variable
+        mask : theano.TensorType
+            Theano variable denoting whether each time step in each
+            sequence in the batch is part of the sequence or not.  If None,
+            then it assumed that all sequences are of the same length.  If
+            not all sequences are of the same length, then it must be
+            supplied as a matrix of shape (n_batch, n_time_steps) where
+            `mask[i, j] = 1` when `j <= (length of sequence i)` and
+            `mask[i, j] = 0` when `j > (length of sequence i)`.
+        """
+
+        # Treat all layers after the first as flattened feature dimensions
+        if input.ndim > 3:
+            input = input.reshape((input.shape[0], input.shape[1],
+                                   T.prod(input.shape[2:])))
+
+        # Because scan iterates over the first dimension we dimshuffle to
+        # (n_time_steps, n_batch, n_features)
+        input = input.dimshuffle(1, 0, 2)
+        seq_len, num_batch, _ = input.shape
+
+        # precompute inputs*W and dimshuffle
+        # Input is provided as (n_batch, n_time_steps, n_features)
+        # W_in is (n_features, 3*num_units). input dot W is then
+        # (n_batch, n_time_steps, 3*num_units). Because scan iterate over the
+        # first dimension we dimshuffle to (n_time_steps, n_batch, n_features)
+        input_dot_w = T.dot(input, self.W_in_to_gates) + self.b_gates
+
+        # input_dow_w is (n_batch, n_time_steps, 2*num_units). We define a
+        # slicing function that extract the input to each GRU gate
+        def slice_w(x, n):
+            return x[:, n*self.num_units:(n+1)*self.num_units]
+
+        # Create single recurrent computation step function
+        # input_dot_W_n is the n'th row of the input dot W multiplication
+        def step(input_dot_w_n, hid_previous, W_hid_to_gates):
```

Comment:
Do you mind making it `input_dot_W_n` for consistency, even though your editor doesn't like it?